.github/workflows/main.yml

```diff
@@ -2,66 +2,193 @@ name: main
+        runs-on: ubuntu-latest
+        steps:
+            - uses: actions/checkout@v4
+            - uses: actions/setup-node@v4
+              with:
+                  node-version: 20
+            - uses: actions/cache/restore@v4
+              with:
+                  path: node_modules
+                  key: ${{ needs.deps.outputs.cache-key }}
+
+            - run: yarn format
+    tests:
+        if: needs.changed-files.outputs.any_changed == 'true'
+        needs: [changed-files, set-matrix, deps, compile]
+        runs-on: ubuntu-latest
+        strategy:
+            matrix:
+                dir: ${{ fromJson(needs.set-matrix.outputs.matrix) }}
+
+        steps:
+            - uses: actions/checkout@v4
+            - uses: actions/setup-node@v4
+              with:
+                  node-version: 20
+            - uses: actions/cache/restore@v4
+              with:
+                  path: node_modules
+                  key: ${{ needs.deps.outputs.cache-key }}
+            - uses: actions/download-artifact@v4
+              with:
+                  name: all-artifacts
+                  path: packages/
 
-            - if: steps.changed-files.outputs.any_changed == 'true'
-              name: Test contracts
-              run: yarn test
+            - if: contains(needs.changed-files.outputs.modified_files, matrix.dir)
+              name: Test
+              run: |
+                  workspace=$(jq -r '.name' packages/${{ matrix.dir }}/package.json)
+                  yarn workspace "$workspace" run test:coverage
 
-            - if: steps.changed-files.outputs.any_changed == 'true' && github.event_name == 'push' && github.ref == 'refs/heads/main'
+            - if: github.event_name == 'push' && github.ref == 'refs/heads/main'
               name: Coveralls
               uses: coverallsapp/github-action@v2
               with:
                   github-token: ${{ secrets.GITHUB_TOKEN }}
+
+    set-matrix:
+        if: needs.changed-files.outputs.any_changed == 'true'
+        needs: changed-files
+        runs-on: ubuntu-latest
+        outputs:
+            matrix: ${{ steps.set-matrix.outputs.matrix }}
+        steps:
+            - uses: actions/checkout@v4
+            - name: Set matrix
+              id: set-matrix
+              run: |
+                  matrix=$(ls -1 packages | jq -Rsc 'split("\n") | map(select(length > 0))')
+                  echo "matrix=$matrix" >> $GITHUB_OUTPUT
+
+    slither:
+        if: needs.changed-files.outputs.any_changed == 'true'
+        needs: [changed-files, set-matrix, deps]
+        runs-on: ubuntu-latest
+        permissions:
+            contents: read
+            security-events: write
+        strategy:
+            matrix:
+                dir: ${{ fromJson(needs.set-matrix.outputs.matrix) }}
+        steps:
+            - uses: actions/checkout@v4
+
+            # FIXME this does not work as a way to restore compilation results for slither job but it does for the compile job ??
```

Comment:
This is not clear. Were you trying to re-use the already generated artifacts here?

Reply:
I thought I could reuse the artifacts generated from the `compile` job yes.
Like I am successfully doing in the `tests` job. see your previous question
https://github.com/privacy-scaling-explorations/zk-kit.solidity/pull/8#discussion_r1654466629

But somehow I had errors with `slither` if using `actions/dowload-artifact` to get the contracts compilation artifacts. So I am compiling again here (not a big deal, just less efficient/ slower).
I can't find an example failing workflow. So let me test again.

Reply:
I see! If it doesn't work it's still fine agree.